### PR TITLE
Breaking: Convert to ESM

### LINKED ISSUE #111: Convert to ESM
I'd like to see if we can convert this package to ESM.

https://blog.sindresorhus.com/hello-modules-d1010b4e777b

That would likely be the final change before the next major release.

### PR DESCRIPTION
**Pull request by bmish**
Fixes #111.

Follows this migration guide: https://blog.sindresorhus.com/hello-modules-d1010b4e777b

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 module.exports = {
     root: true,
+    parserOptions: {
+        ecmaVersion: "2020"
```
Comment (nzakas) [inline .eslintrc.cjs:6]:
```suggestion
        ecmaVersion: 2020
```

File: app/index.js [hunk 2]
```diff
@@ -4,28 +4,27 @@
...
 
 //------------------------------------------------------------------------------
 // Requirements
 //------------------------------------------------------------------------------
 
-const Generator = require("yeoman-generator");
+import Generator from "yeoman-generator";
+import RuleGenerator from "../rule/index.js";
+import PluginGenerator from "../plugin/index.js";
+import { fileURLToPath } from "node:url"; // eslint-disable-line node/no-missing-import -- https://github.com/mysticatea/eslint-plugin-node/issues/275
```
Comment (nzakas) [inline app/index.js:15]: Just double checking that the “node:” prefix is supported in all of the Node.JS versions we support? (12, 14, 16)?

File: plugin/index.js [hunk 3]
```diff
@@ -4,28 +4,24 @@
...
 
 //------------------------------------------------------------------------------
 // Requirements
 //------------------------------------------------------------------------------
 
-const Generator = require("yeoman-generator");
-const { mkdirSync } = require("fs");
-const validators = require("../lib/validators");
+import Generator from "yeoman-generator";
+import { mkdirSync } from "node:fs"; // eslint-disable-line node/no-missing-import -- https://github.com/mysticatea/eslint-plugin-node/issues/275
```
Comment (nzakas) [inline plugin/index.js:13]: Can we move the comment to shorten line length?

```suggestion
// eslint-disable-next-line node/no-missing-import -- https://github.com/mysticatea/eslint-plugin-node/issues/275
import { mkdirSync } from "node:fs";
```

File: tests/test-load.js [hunk 4]
```diff
@@ -3,13 +3,13 @@
  * @author Nicholas C. Zakas
  */
 
-"use strict";
+import assert from "node:assert"; // eslint-disable-line node/no-missing-import -- https://github.com/mysticatea/eslint-plugin-node/issues/275
```
Comment (nzakas) [inline tests/test-load.js:6]:
```suggestion
// eslint-disable-next-line node/no-missing-import -- https://github.com/mysticatea/eslint-plugin-node/issues/275
import assert from "node:assert"; 
```

File: tests/test-load.js [hunk 5]
```diff
@@ -3,13 +3,13 @@
...
 
-const assert = require("assert");
+const importFresh = async modulePath => import(`${modulePath}?x=${new Date()}`); // eslint-disable-line func-style, node/no-unsupported-features/es-syntax -- https://github.com/mysticatea/eslint-plugin-node/issues/250
```
Comment (nzakas) [inline tests/test-load.js:8]:
```suggestion
// eslint-disable-line func-style, node/no-unsupported-features/es-syntax -- https://github.com/mysticatea/eslint-plugin-node/issues/250
const importFresh = async modulePath => import(`${modulePath}?x=${new Date()}`);
```

**Review by nzakas — CHANGES_REQUESTED** [review]
Overall looks good, just a couple of minor things.

File: app/index.js (hunk 2 above)
Comment (bmish) [reply app/index.js:15]: Yep. Unfortunately, eslint-plugin-node hasn't released a new version with the fix yet.

File: plugin/index.js (hunk 3 above)
Comment (bmish) [reply plugin/index.js]: Updated.

File: tests/test-load.js (hunk 5 above)
Comment (bmish) [reply tests/test-load.js:9]: Updated.

File: tests/test-load.js (hunk 4 above)
Comment (bmish) [reply tests/test-load.js]: Updated.

**Review by nzakas — APPROVED** [review]
LGTM. Thanks!

File: tests/test-load.js (hunk 5 above)
Comment (btmills) [reply tests/test-load.js:9]: Looks like this one didn't get applied with the rest of the batch. The earlier suggestion was missing `next-`, so adding another one here for convenience:

```suggestion
// eslint-disable-next-line func-style, node/no-unsupported-features/es-syntax -- https://github.com/mysticatea/eslint-plugin-node/issues/250
const importFresh = async modulePath => import(`${modulePath}?x=${new Date()}`);
```

**Review by btmills — COMMENTED** [review]
One more line length suggestion, then LGTM. Thanks for all the updates to this repository!

**Comment by nzakas** [comment]
@bmish please wait for someone on the TSC to merge changes other than chores, bugs, and docs. We still need to review other changes. You can always request that something be merged if you think we forgot, but we do have a set process we need to follow.

**Comment by bmish** [comment]
Ah, my bad, sorry about that. I'm aware of that guidance but it slipped my mind here as I was trying to make sure everything was in order for the upcoming release.

**Comment by nzakas** [comment]
No worries. We only stick to a set schedule for the eslint package itself, the others are more ad-how because they have fewer changes. Do you think generator-eslint is ready for a release now? If so I can tee up a release for next week.

**Comment by bmish** [comment]
Yes I would say we're ready for a major release. Someone could test it out locally just to double-check everything looks good.

**Comment by nzakas** [comment]
Yup, I’ll do that and schedule a release.

**Pull request merged**